server/service/global_policies.go

```diff
@@ -550,3 +557,124 @@ func (svc *Service) ApplyPolicySpecs(ctx context.Context, policies []*fleet.Poli
 	}
 	return nil
 }
+
+/////////////////////////////////////////////////////////////////////////////////
+// Autofill
+/////////////////////////////////////////////////////////////////////////////////
+
+type autofillPoliciesRequest struct {
+	SQL string `json:"sql"`
+}
+
+type autofillPoliciesResponse struct {
+	Description string `json:"description"`
+	Resolution  string `json:"resolution"`
+	Err         error  `json:"error,omitempty"`
+}
+
+func (a autofillPoliciesResponse) error() error {
+	return a.Err
+}
+
+func autofillPoliciesEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*autofillPoliciesRequest)
+	description, resolution, err := svc.AutofillPolicySql(ctx, req.SQL)
+	return autofillPoliciesResponse{Description: description, Resolution: resolution, Err: err}, nil
+}
+
+// Exposing external URL and timeout for testing purposes
+var getHumanInterpretationFromOsquerySqlUrl = "https://fleetdm.com/api/v1/get-human-interpretation-from-osquery-sql"
+var getHumanInterpretationFromOsquerySqlTimeout = 30 * time.Second
+
+type AutofillError struct {
+	Message     string
+	InternalErr error
+}
+
+// Error implements the error interface.
+func (e AutofillError) Error() string {
+	return e.Message
+}
+
+// StatusCode implements the kithttp.StatusCoder interface.
+func (e AutofillError) StatusCode() int {
+	return http.StatusUnprocessableEntity
+}
+
+func (e AutofillError) Internal() string {
+	if e.InternalErr == nil {
+		return ""
+	}
+	return e.InternalErr.Error()
+}
+
+func (svc *Service) AutofillPolicySql(ctx context.Context, sql string) (description string, resolution string, err error) {
+	// We expect that only users with policy write permissions will autofill policies.
+	if err = svc.authz.Authorize(ctx, &fleet.Policy{}, fleet.ActionWrite); err != nil {
+		return "", "", err
+	}
+
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return "", "", err
+	}
+	if appConfig.ServerSettings.AIFeaturesDisabled {
+		return "", "", ctxerr.Wrap(
+			ctx, &fleet.BadRequestError{
+				Message: "AI features are disabled (server_settings.ai_features_disabled)",
+			},
+		)
+	}
+
+	sql = strings.TrimSpace(sql)
+	if sql == "" {
+		return "", "", ctxerr.Wrap(ctx, &fleet.BadRequestError{Message: "'sql' cannot be empty"})
+	}
+
+	// Using a timeout smaller than the Fleet server's WriteTimeout
+	client := fleethttp.NewClient(fleethttp.WithTimeout(getHumanInterpretationFromOsquerySqlTimeout))
+	reqBody := fmt.Sprintf(`{"sql": "%s"}`, sql)
```

Comment:
Should double quotes in the SQL be escaped?

Reply:
Yes. Switched to using `json.Marshal`